src/meta/src/barrier/recovery.rs

```diff
@@ -220,45 +220,101 @@ where
+                .retain(|_, to| expired_workers.contains(to) || info.actor_map.contains_key(to));
+            cached_plan.parallel_unit_plan.retain(|_, to| {
+                expired_workers.contains(&to.worker_node_id)
+                    || info.actor_map.contains_key(&to.worker_node_id)
+            });
+
+            expired_workers.retain(|id| !cached_plan.worker_plan.contains_key(id));
+        }
+
+        if expired_workers.is_empty() {
+            // all expired workers are already in migration plan.
+            debug!("all expired workers are already in migration plan.");
+            return Ok(cached_plan);
+        }
+        debug!("got expired workers {:#?}", expired_workers);
+        let mut expired_workers = expired_workers.into_iter().collect_vec();
+        let all_worker_parallel_units = self.fragment_manager.all_worker_parallel_units().await;
+
         let start = Instant::now();
-        while cur < expired_workers.len() {
+        // if expire workers are not empty, should wait for newly joined worker.
+        'discovery: while !expired_workers.is_empty() {
             let current_nodes = self
                 .cluster_manager
                 .list_active_streaming_compute_nodes()
                 .await;
             let new_nodes = current_nodes
                 .into_iter()
                 .filter(|node| {
-                    !info.actor_map.contains_key(&node.id) && !node_map.contains_key(&node.id)
+                    !info.actor_map.contains_key(&node.id)
+                        && !cached_plan.worker_plan.values().contains(&node.id)
                 })
                 .collect_vec();
+
             for new_node in new_nodes {
-                let actors = info.actor_map.get(&expired_workers[cur]).unwrap();
-                for actor in actors {
-                    migrate_map.insert(*actor, new_node.id);
-                }
-                cur += 1;
-                debug!(
-                    "new worker joined: {}, migrate process ({}/{})",
-                    new_node.id,
-                    cur,
-                    expired_workers.len()
-                );
-                node_map.insert(new_node.id, new_node);
-                if cur == expired_workers.len() {
-                    return (migrate_map, node_map);
+                if let Some(from) = expired_workers.pop() {
+                    debug!(
+                        "new worker joined, plan to migrate from worker {} to {}",
+                        from, new_node.id
+                    );
+                    cached_plan.worker_plan.insert(from, new_node.id);
+                    assert!(all_worker_parallel_units.contains_key(&from));
+                    let from_parallel_units = all_worker_parallel_units.get(&from).unwrap();
+                    // todo: remove it and migrate actors only based on parallel unit mapping.
+                    assert!(from_parallel_units.len() <= new_node.parallel_units.len());
```

Comment:
This assertion can fail.

Reply:
Yes, that's the reason that I added a todo here. Since we're using same pu config for all CNs, it works at least for now. Besides, to prevent incorrect migration plan to be generated in some other deployment if any, I added this assertion here. Anyway, it will be resolved together with the todo.